datalad/support/gitrepo.py

```diff
@@ -2355,17 +2358,67 @@ def get_submodules_(self, paths=None):
             return
 
         modinfo = self._parse_gitmodules()
-        for path, props in self.get_content_info(
-                paths=paths,
-                ref=None,
-                untracked='no').items():
-            if props.get('type', None) != 'dataset':
+        if not modinfo:
+            # we exit early, if there is nothing on record (even though
+            # a .gitmodules file exists).
+            # without this conditional exit, we would be able to discover
+            # subprojects even when they are not recorded in .gitmodules,
+            # but at the cost of running an unconstrained ls-files call
+            # below
+            return
+
+        # taken from 3.9's pathlib, can be removed once the minimally
+        # supported python version
+        def is_relative_to(self, *other):
+            """Return True if the path is relative to another path or False.
+            """
+            try:
+                self.relative_to(*other)
+                return True
+            except ValueError:
+                return False
+
+        if paths:
+            # ease comparison
+            paths = [self.pathobj / p for p in paths]
+            # contrain the report by the given paths
+            modinfo = {
+                # modpath is absolute
+                modpath: modprobs
+                for modpath, modprobs in modinfo.items()
+                # is_relative_to() also match equal paths
+                if any(is_relative_to(modpath, p) for p in paths)
+            }
+        for r in self.call_git_items_(
+            ['ls-files', '--stage', '-z'],
+            sep='\0',
+            files=[str(p.relative_to(self.pathobj)) for p in modinfo.keys()],
+            read_only=True,
+            keep_ends=True,
+        ):
+            if not r.startswith('160000'):
                 # make sure this method never talks about non-dataset
                 # content
                 continue
-            props["path"] = path
-            props.update(modinfo.get(path, {}))
-            yield props
+            props, rpath = r.split('\t')
+            mode, gitsha, stage = props.split(' ')
+            if stage not in ('0', '2'):
+                # we either have non-merge situation, or a simple merge
+                # situation (i.e. stage=0). the reported gitsha always
+                # matches what we have locally.
+                # or we are in a three-way merge, in which case stage=2
+                # is what we want to report, because it matches the
+                # current HEAD (see git-read-tree manpage).
+                # there is either a stage 2 or stage 0, never both
+                continue
```

Comment:
Not familiar much with stages, but it is not clear why we would skip reporting submodule if it is in oms other `stage?  e.g. would it then be present in `stage` `0` and thus still be yielded below? 
is there a test which would demonstrate that may be?

Reply:
Yes, without this conditional. Our merge conflict tests in `update` fail.

Reply:
https://git-scm.com/docs/git-read-tree#_3_way_merge
I read it like this: if stages 1 and 3 are present, they belong to the common ancestor and the remote submodule state - stage 2 is "our" submodule state. Stages 1-3 won't exist if stage 0 exists, and if stage 0 exists, its always the state of "our" submodule.

Reply:
thank you @mih and @adswa !